### DISCUSSION TITLE
doNd measurements of a parameter with setpoints fail, whereas measurements with context manager work

### DISCUSSION THREAD

**Discussion by simonzihlmann** [microsoft/Qcodes#4331] [Q&A] [answered]
Recently we run into a strange problem while writing some helper instrument/parameters to get the data from the Zurich UHF data acquisition module into some nice qcodes parameters (parameters with setpoints mostly). Namely, the measurement of these parameters fails when using doNd functions. At the same time we can though perform measurements of the same parameters with the general measurement context manager.

Below are some code snippets illustrating the behaviour.

Trying with a doNd function:
```python
do0d(daqh.dev2226_demods_0_sample_r.average_timetrace, measurement_name='test')

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [46], in <cell line: 1>()
----> 1 do0d(daqh.dev2226_demods_0_sample_r.average_timetrace, measurement_name='test')

File ~\Miniconda3\lib\site-packages\qcodes\utils\dataset\doNd.py:172, in do0d(write_period, measurement_name, exp, do_plot, use_threads, log_info, *param_meas)
    169 _register_parameters(meas, param_meas, shapes=shapes)
    170 _set_write_period(meas, write_period)
--> 172 with meas.run() as datasaver:
    173     datasaver.add_result(*process_params_meas(param_meas, use_threads=use_threads))
    174     dataset = datasaver.dataset

File ~\Miniconda3\lib\site-packages\qcodes\dataset\measurements.py:601, in Runner.__enter__(self)
    598 else:
    599     snapshot = {}
--> 601 self.ds.prepare(
    602     snapshot=snapshot,
    603     interdeps=self._interdependencies,
    604     write_in_background=self._write_in_background,
    605     shapes=self._shapes,
    606     parent_datasets=self._parent_datasets,
    607 )
    609 # register all subscribers
    610 if isinstance(self.ds, DataSet):

File ~\Miniconda3\lib\site-packages\qcodes\dataset\data_set.py:331, in DataSet.prepare(self, snapshot, interdeps, shapes, parent_datasets, write_in_background)
    329 links = [Link(head=self.guid, **pdict) for pdict in parent_datasets]
    330 self.parent_dataset_links = links
--> 331 self.mark_started(start_bg_writer=write_in_background)

File ~\Miniconda3\lib\site-packages\qcodes\dataset\data_set.py:644, in DataSet.mark_started(self, start_bg_writer)
    633 """
    634 Mark this :class:`.DataSet` as started. A :class:`.DataSet` that has been started can not
    635 have its parameters modified.
   (...)
    641         database in a separate thread.
    642 """
    643 if not self._started:
--> 644     self._perform_start_actions(start_bg_writer=start_bg_writer)
    645     self._started = True

File ~\Miniconda3\lib\site-packages\qcodes\dataset\data_set.py:658, in DataSet._perform_start_actions(self, start_bg_writer)
    653 for spec in paramspecs:
    654     add_parameter(
    655         spec, conn=self.conn, run_id=self.run_id, insert_into_results_table=True
    656     )
--> 658 desc_str = serial.to_json_for_storage(self.description)
    660 update_run_description(self.conn, self.run_id, desc_str)
    662 set_run_timestamp(self.conn, self.run_id)

File ~\Miniconda3\lib\site-packages\qcodes\dataset\descriptions\versioning\serialization.py:132, in to_json_for_storage(desc)
    127 def to_json_for_storage(desc: current.RunDescriber) -> str:
    128     """
    129     Serialize the given RunDescriber to JSON as a RunDescriber of the
    130     version for storage
    131     """
--> 132     return json.dumps(to_dict_for_storage(desc))

File ~\Miniconda3\lib\json\__init__.py:231, in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    226 # cached encoder
    227 if (not skipkeys and ensure_ascii and
    228     check_circular and allow_nan and
    229     cls is None and indent is None and separators is None and
    230     default is None and not sort_keys and not kw):
--> 231     return _default_encoder.encode(obj)
    232 if cls is None:
    233     cls = JSONEncoder

File ~\Miniconda3\lib\json\encoder.py:199, in JSONEncoder.encode(self, o)
    195         return encode_basestring(o)
    196 # This doesn't pass the iterator directly to ''.join() because the
    197 # exceptions aren't as detailed.  The list call should be roughly
    198 # equivalent to the PySequence_Fast that ''.join() would do.
--> 199 chunks = self.iterencode(o, _one_shot=True)
    200 if not isinstance(chunks, (list, tuple)):
    201     chunks = list(chunks)

File ~\Miniconda3\lib\json\encoder.py:257, in JSONEncoder.iterencode(self, o, _one_shot)
    252 else:
    253     _iterencode = _make_iterencode(
    254         markers, self.default, _encoder, self.indent, floatstr,
    255         self.key_separator, self.item_separator, self.sort_keys,
    256         self.skipkeys, _one_shot)
--> 257 return _iterencode(o, 0)

File ~\Miniconda3\lib\json\encoder.py:179, in JSONEncoder.default(self, o)
    160 def default(self, o):
    161     """Implement this method in a subclass such that it returns
    162     a serializable object for ``o``, or calls the base implementation
    163     (to raise a ``TypeError``).
   (...)
    177 
    178     """
--> 179     raise TypeError(f'Object of type {o.__class__.__name__} '
    180                     f'is not JSON serializable')

TypeError: Object of type uintc is not JSON serializable
```

Same with the context manager:
```python
context_meas = Measurement(exp=exp, station=station, name='context_example')

# register parameter
context_meas.register_parameter(daqh.dev2226_demods_0_sample_r.average_timetrace)

with context_meas.run() as datasaver:
    datasaver.add_result((daqh.dev2226_demods_0_sample_r.average_timetrace, daqh.dev2226_demods_0_sample_r.average_timetrace()))

    dataset = datasaver.dataset

# plot dataset
plot_dataset(dataset)
```

The `daqh.dev2226_demods_0_sample_r.average_timetrace` is a parameter with setpoints, that is attached to a helper instrument `(daqh)`.

Does anyone know why the doNd does not work? What is the difference between using doNd and the context manager? Does the context manager also generate the same RunDescriber?

Any help/comment is highly appreciated.

Qcodes version used: '0.33.0'

**Comment by astafan8** [comment]
@simonzihlmann this is a very interesting one. It seems like some number that is part of the rundescriber, likely the shape of the parameter, is not of a json-able type (the ``uintc``) type, so would be nice to know which one and why.

For the better understanding of the issue, could you try running the second snippet of the code and after `context_meas.register_parameter(...` call execute this print ``print(context_meas._interdeps), print(context_meas._shapes)`` and paste here what you got?

and perhaps the code for the parameter with setpoints `daqh.dev2226_demods_0_sample_r.average_timetrace` is available somewhere? perhaps you can give a link or a paste the relevant code here?

**Reply by simonzihlmann** [reply]
Here is the requested print.

code:
```python
context_meas = Measurement(exp=exp, station=station, name='context_example')

# register parameter
context_meas.register_parameter(daqh.dev2226_demods_0_sample_r.average_timetrace)

# print shapes
print(context_meas._interdeps), print(context_meas._shapes)
```

print:
```python
InterDependencies_(dependencies={ParamSpecBase('guv_dev2226_demods_0_sample_r_average_timetrace', 'array', 'averaged timetrace over the rows', 'V'): (ParamSpecBase('guv_dev2226_demods_0_sample_r_time_axis', 'array', 'time', 's'),)}, inferences={}, standalones=frozenset())
None
(None, None)
```

**Comment by simonzihlmann** [comment]
Thanks for the rapid response @astafan8 
I will post `print(context_meas._interdeps), print(context_meas._shapes)` here once I get access to the experiment (this afternoon).

The code is not published anywhere, but I can post the relevant sections here.

Definition of the paramter
```python
        # 1D average timetrace over the different rows
        self.add_parameter(name='average_timetrace',
                           label='averaged timetrace over the rows',
                           unit=self._unit,
                           get_cmd=self._get_average_timetrace,
                           setpoints=(self.time_axis,),
                           parameter_class=ParameterWithSetpoints,
                           vals=Arrays(shape=(self._cols,)),
                           snapshot_exclude=False
                          )
```

updating the shape of the parameter depending on the data read from the data acquisition module
```python
    def update_param(self) -> None:
        """
        Updates the shapes of the parameters. This is called after getting
        the helper parameter.
        """
        #print('update_param')
        for _, parameter in self.parameters.items():
            if parameter.name in ['time_axis', 'average_timetrace']:
                try:
                    #print('shape of {}: {}'.format(parameter.name, self._shape_time))
                    parameter.vals=Arrays(shape = (self._cols,))
                except AttributeError:
                    print('Error: could not update shape of time_axis or average_timetrace')
                    pass
            if parameter.name in ['timestamp_rows', 'timetrace_rows']:
                try:
                    #print('shape of {}: {}'.format(parameter.name, self._shape_data))
                    parameter.vals=Arrays(shape = self._shape_data)
                except AttributeError:
                    print('Error: could not update shape of timestamp_rows or timetrace_rows')
                    pass
            if parameter.name=='rows_axis':
                try:
                    #print('shape of {}: {}'.format(parameter.name, (self._rows,)))
                    parameter.vals=Arrays(shape = (self._rows,))
                except AttributeError:
                    print('Error: could not update shape of rows_axis')
                    pass
```

populating the _shape_xxx attributes derived from the data.
```python
            self.instrument.submodules[ch_name]._shape_data = (
                                                self.instrument.submodules[ch_name]._channel_results[0]['header']['gridrows'][0],
                                                self.instrument.submodules[ch_name]._channel_results[0]['header']['gridcols'][0]
                                                )
            # self.instrument.submodules[ch_name]._shape_time = (self.instrument.submodules[ch_name]._results.shape[1], )
            self.instrument.submodules[ch_name]._rows = self.instrument.submodules[ch_name]._shape_data[0]
            self.instrument.submodules[ch_name]._cols = self.instrument.submodules[ch_name]._shape_data[1]
```

I have attached the complete code in a zip file.
The general idea of the code is the following. There is an instrument `DAQHelper` that exposes qcodes-parameters to the end user for each node subscribed to in the Zurich UHF data acquisition module. A parameter `DAQHelperParameter` interacts with the hardware and prepares the data for later use. Then for each subscribed node, a Channel is created `DAQSignalChannel,` to which all qcodes parameters are bound to (e.g. `average_timetrace`). Some of them contain just the raw data, some others will contain processed data (e.g. the average of a time trace). Then the user can easily interact wit the qcodes parameter and does not need to worry about how to get the data in the first place.

[DAQHelper_new_driver.zip](https://github.com/QCoDeS/Qcodes/files/9019037/DAQHelper_new_driver.zip)

**Comment by jenshnielsen** [comment]
What I can gather from your error is that the shape is a `uintc `  which is a unsigned c integer. I would guess this is probably the version that is defined in numpy https://numpy.org/devdocs/reference/arrays.scalars.html#numpy.uintc

However I cannot reproduce any issue with serializing that type so perhaps its not the type defined in numpy?

```
import json
try:
    from qcodes.utils import NumpyJSONEncoder # master
except ImportError:
    from qcodes.utils.helpers import NumpyJSONEncoder

json.dumps({"a": np.uintc(1)}, cls=NumpyJSONEncoder)
```

**Comment by astafan8** [comment, marked as answer]
@jenshnielsen indeed, the serialization works with qcodes NumpyJSONEncoder but it does not without the class, and note that the class is not used for json-serialize-ing the rundescriber.

so here's the minimal example:
```python
import json

import numpy as np

json.dumps({"a": np.uintc(1)})
```

```bash
Traceback (most recent call last):
  File "C:\tools\miniconda3\envs\qchar\lib\site-packages\IPython\core\interactiveshell.py", line 3444, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-1ac48cfc489e>", line 1, in <module>
    json.dumps({"a": np.uintc(1)})
  File "C:\tools\miniconda3\envs\qchar\lib\json\__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "C:\tools\miniconda3\envs\qchar\lib\json\encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "C:\tools\miniconda3\envs\qchar\lib\json\encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "C:\tools\miniconda3\envs\qchar\lib\json\encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type uintc is not JSON serializable

```

-------------

@simonzihlmann thanks for the details! I think @jenshnielsen is spot on, likely the values `self.instrument.submodules[ch_name]._channel_results[0]['header']['gridrows'][0]` and so on are of type `np.uintc`, so the fix is to explicitly convert them to python int by calling `int` on them, e.g. `int( self.instrument.submodules[ch_name]._channel_results[0]['header']['gridrows'][0] )`. Could you try that after you do the prints that asked for above?

--------------

@simonzihlmann to your more general questions

> What is the difference between using doNd and the context manager? Does the context manager also generate the same RunDescriber?

I refer you to the qcodes documentation: https://qcodes.github.io/Qcodes/examples/15_minutes_to_QCoDeS.html#Using-the-doNd-multi-dimensional-measurement-utility and https://qcodes.github.io/Qcodes/examples/DataSet/Using_doNd_functions_in_comparison_to_Measurement_context_manager_for_performing_measurements.html.

and yes, context manager generates the same RunDescriber because dond uses the context manager inside itself.

**Reply by simonzihlmann** [reply]
Thanks @astafan8 and @jenshnielsen for your fast reply and help.

I indeed confirm that the type of `self.instrument.submodules[ch_name]._channel_results[0]['header']['gridrows'][0]` is `numpy.uintc`

I also confirm that do0d works as expected after converting it explicitly to `int`.

But what I still do not understand is why the context manager worked whereas the doNd failed. Did I miss something on how I used the context manager? I will look at the docs but I think I already looked at it many times.

Thanks again!

**Reply by jenshnielsen** [reply]
> But what I still do not understand is why the context manager worked whereas the doNd failed. Did I miss something on how I used the context manager? I will look at the docs but I think I already looked at it many times.

I honersly don't know Either. I would expect both of them to fail

**Reply by jenshnielsen** [reply]
Actually its probably because do_nd sets the shape of the measurement before starting it. I bet if you do that manually you will see the same issue https://qcodes.github.io/Qcodes/examples/DataSet/Performing-measurements-using-qcodes-parameters-and-dataset.html#Specifying-shape-of-measurement
